File: Python/Maximum Subarray.py
"""
Find the contiguous subarray within an array (containing at least one number) which has the largest sum.
For example, given the array [−2,1,−3,4,−1,2,1,−5,4],
the contiguous subarray [4,−1,2,1] has the largest sum = 6.
"""

class Solution:
    # @param A, a list of integers
    # @return an integer
    def maxSubArray(self, A):
        maxSum = A[0]
        maxEnd = A[0]
        for i in range(1,len(A)):
            maxEnd = max(maxEnd+A[i],A[i])
            maxSum = max(maxSum,maxEnd)
        return maxSum


# Greedy Algorithm
class Solution:
    # @param A, a list of integers
    # @return an integer
    def maxSubArray(self, A):
        Max = A[0]
        current = 0
        
        for num in A:
            current += num
            Max = max(Max,current)
            current = max(0,current)
        return Max

File: Java/Maximum Subarray.java
/*
Find the contiguous subarray within an array (containing at least one number) that has the largest sum.
For example, given the array [2, 1, –3, 4, –1, 2, 1, –5, 4], The contiguous array [4, –1, 2, 1] has the largest sum = 6.
*/

//O(n) time and O(1) space - Dynamic Programming:
public class Solution {
    public int maxSubArray(int[] nums) {
        if(nums.length == 0) return 0;
        int maxEndingHere = nums[0];
        int maxSoFar = nums[0];
        for(int i = 1;i<nums.length;i++){
            maxEndingHere = Math.max(maxEndingHere+nums[i],nums[i]);
            maxSoFar = Math.max(maxSoFar,maxEndingHere);
        }
        return maxSoFar;
    }
}


//O(n log n) runtime, O(log n) stack space – Divide and Conquer:
public class Solution {
    public int maxSubArray(int[] nums) {
        return maxSubArrayHelper(nums,0,nums.length-1);
    }
    
    public int maxSubArrayHelper(int[] nums,int left,int right){
        if(left>right) return Integer.MIN_VALUE;
        int middle = left + (right - left)/2;
        int leftAnswer = maxSubArrayHelper(nums,left,middle-1);
        int rightAnswer = maxSubArrayHelper(nums,middle+1,right);
        int leftMaxSum = 0, sum = 0;
        for(int i = middle - 1;i>=left;i--){
            sum+=nums[i];
            leftMaxSum = Math.max(sum,leftMaxSum);
        }
        int rightMaxSum = 0;
        sum = 0;
        for(int i = middle + 1;i<=right;i++){
            sum+=nums[i];
            rightMaxSum = Math.max(sum,rightMaxSum);
        }
        return Math.max(leftMaxSum + nums[middle] + rightMaxSum,Math.max(leftAnswer,rightAnswer));
    }
}